profiling/src/collections/identifiable/string_table.rs

```diff
@@ -0,0 +1,266 @@
+            vec: Default::default(),
+            map: Default::default(),
+        });
+
+        let mut s = Self { inner };
+        // string tables always have the empty string at 0.
+        let (_id, _inserted) = s.insert_full("");
+        debug_assert!(_id == StringId::ZERO);
+        debug_assert!(_inserted);
+        s
+    }
+
+    #[allow(unused)]
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.inner.with_dependent(|table| table.vec.len())
+    }
+
+    #[allow(unused)]
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Inserts the string into the table, if it did not already exist. The id
+    /// of the string is returned.
+    ///
+    /// # Panics
+    /// Panics if a new string needs to be inserted but the offset of the new
+    /// string doesn't fit into a [StringId].
+    #[inline]
+    pub fn insert(&mut self, str: &str) -> StringId {
+        self.insert_full(str).0
+    }
+
+    /// Inserts the string into the table, if it did not already exist. The id
+    /// of the string is returned, along with whether the string was inserted.
+    ///
+    /// # Panics
+    /// Panics if a new string needs to be inserted but the offset of the new
+    /// string doesn't fit into a [StringId].
+    #[inline]
+    pub fn insert_full(&mut self, str: &str) -> (StringId, bool) {
+        // For performance, delay converting the &str to a String until after
+        // it has been determined to not exist in the set. This avoids
+        // temporary allocations.
+        self.inner
+            .with_dependent_mut(|table| match table.map.get(str) {
+                None => {
+                    let id = StringId::from_offset(table.vec.len());
+                    let bumped_str = table.arena.alloc_str(str);
+
+                    table.vec.push(bumped_str);
+                    table.map.insert(bumped_str, id);
+                    assert_eq!(table.vec.len(), table.map.len());
+                    (id, true)
+                }
+                Some(id) => (*id, false),
+            })
+    }
+
+    /// Gets the string associated with the id.
+    ///
+    /// # Panics
+    /// Panics if the [StringId] doesn't exist in the table.
+    #[inline]
+    pub fn get_id(&self, id: StringId) -> &str {
+        self.inner.with_dependent(|table| {
+            let offset = id.to_offset();
+            match table.vec.get(offset) {
+                Some(str) => str,
+                None => panic!("expected string id {offset} to exist in the string table"),
+            }
+        })
+    }
+
+    #[cfg(test)]
+    #[allow(unused)]
+    #[inline]
+    pub fn get_range(&self, range: Range<usize>) -> &[&str] {
+        self.inner.with_dependent(|table| &table.vec[range])
+    }
+
+    /// Returns an iterator over the strings in the table. The items are
+    /// returned in the order they were inserted, matching the [StringId]s.
+    pub fn iter(&self) -> impl Iterator<Item = &str> {
```

Comment:
Not necessary, and might be tricky with lifetimes but this got me wondering if we can have an `into_iter` implementation that drops the underlying storage when the iteration is done.

Reply:
It's not worth it, IMO, because we'd have to make another self-referencing struct, and it doesn't actually buy us anything (end result is the same).